### PR TITLE
[mac] enable Beacon Payload in outgoing beacons

### PR DESCRIPTION
**Pull request by hemanth-silabs**
Supplement PR to [7736](https://github.com/openthread/openthread/pull/7736)
This PR enables beacon payloads on outgoing beacons to prevent backward compatibility issues with existing devices in the market.

### CODE REVIEW

**Comment by hemanth-silabs** [comment]
> > > I can't see how this PR fixes backward compatibility issues. There are four scenarios:
> > > 
> > > 1. **Old joiner and old network**
> > >    Beacons contain payload and joiner uses it for commissioning purposes. MLE Discovery can be used as well.
> > > 2. **Old joiner and new network**
> > >    Backwards compatibility is broken since the joiner expects beacons payload which is not present. But this PR doesn't fix it.
> > > 3. **New joiner and old network**
> > >    Beacons contain payload but joiner has no code to parse it. Anyway, since it's a new development the joiner can be adapted for using MLE Discovery instead.
> > > 4. **New joiner and new network**
> > >    No payload in beacons, MLE Discovery must be used.
> > > 
> > > Am I missing something?
> > 
> > 
> > @edmont Thank you for your response. The scenarios you list are correct, and this PR tries to mitigate an issue close to 3rd scenario that you list. We have an issue with a joiner not being able to parse the beacons from old network.
> > The issue is that if the joiner is not a new development but instead has an established business logic and application state machine, and just wants to update the stack, the joining won't work. It would be troublesome to change the logic now and when we try to upgrade the stack we see that the joining won't work anymore.
> 
> @hemanth-silabs is this issue related to a device that doesn't actually use meshcop joining? I ask because I discovered a commercially available Thread device that I know uses Silabs and can not attach to my border router with the beacon payload changes. If so, this set of changes don't appear to actually resolve it (I built with these commits and still saw the issue).

@jwhui @abtink @edmont Please have a look at this PR. This Pr enables beacon payload in beacons to avid backward compatibility issues for old devices trying to join new networks.

This is the conversation we had on the previous PR.

**Comment by hemanth-silabs** [comment]
@romacdon @suveshpratapa @gabekassel This PR is to address the other backward compatibility issues we were seeing on existing devices in the market trying to join new networks. Changes are disabled by default and can be enabled by passing the that **`OPENTHREAD_CONFIG_MAC_BEACON_PAYLOAD_ENABLE=1`** while building.

**Comment by romacdon** [comment]
@gabekassel it would be great if you could try out the changes in this PR and verify that it fixes the issue you are experiencing.

**Review by suveshpratapa — APPROVED** [review]

**Comment by jwhui** [comment]
@hemanth-silabs , can you help rebase this PR to fix some of the CI failures?

**Comment by hemanth-silabs** [comment]
> @hemanth-silabs , can you help rebase this PR to fix some of the CI failures?

Thanks @jwhui. Just did a rebase hopefully this fixes the failures.

**Comment by gabekassel** [comment]
> @gabekassel it would be great if you could try out the changes in this PR and verify that it fixes the issue you are experiencing.

@romacdon I built [this](https://github.com/openthread/ot-br-posix/compare/main...gabekassel:gkass/beacon-payload-revert?expand=1) with `-DOPENTHREAD_CONFIG_MAC_BEACON_PAYLOAD_ENABLE=1` defined in my cmake args and still see failure with the commercial device in question.

**Comment by romacdon** [comment]
> > @gabekassel it would be great if you could try out the changes in this PR and verify that it fixes the issue you are experiencing.
> 
> @romacdon I built [this](https://github.com/openthread/ot-br-posix/compare/main...gabekassel:gkass/beacon-payload-revert?expand=1) with `-DOPENTHREAD_CONFIG_MAC_BEACON_PAYLOAD_ENABLE=1` defined in my cmake args and still see failure with the commercial device in question.

Thanks @gabekassel. Can you provide a packet trace so that @hemanth-silabs and @suveshpratapa can debug it.

**Comment by gabekassel** [comment]
> > > @gabekassel it would be great if you could try out the changes in this PR and verify that it fixes the issue you are experiencing.
> > 
> > 
> > @romacdon I built [this](https://github.com/openthread/ot-br-posix/compare/main...gabekassel:gkass/beacon-payload-revert?expand=1) with `-DOPENTHREAD_CONFIG_MAC_BEACON_PAYLOAD_ENABLE=1` defined in my cmake args and still see failure with the commercial device in question.
> 
> Thanks @gabekassel. Can you provide a packet trace so that @hemanth-silabs and @suveshpratapa can debug it.

Attached, though not sure I'm capturing everything desired. This is wireshark using pyspinel setup with the network key. On the OTBR logging side, I just see continual beacon requests. 

Screen Shot 2022-05-27 at 12 40 19 PM

File: src/core/config/mac.h
Comment (romacdon) [inline src/core/config/mac.h:510]: I think we should actually default this to 1.  Without this functionality new builds will not work with deployed devices.  @jwhui @abtink What do you think?

**Review by romacdon — CHANGES_REQUESTED** [review]

File: src/core/config/mac.h
Comment (gabekassel) [reply src/core/config/mac.h:510]: I agree. This is a footgun if defaulted off.

Comment (jwhui) [reply src/core/config/mac.h:510]: Yes, this should default to 1.

**Comment by romacdon** [comment]
> > > > @gabekassel it would be great if you could try out the changes in this PR and verify that it fixes the issue you are experiencing.
> > > 
> > > 
> > > @romacdon I built [this](https://github.com/openthread/ot-br-posix/compare/main...gabekassel:gkass/beacon-payload-revert?expand=1) with `-DOPENTHREAD_CONFIG_MAC_BEACON_PAYLOAD_ENABLE=1` defined in my cmake args and still see failure with the commercial device in question.
> > 
> > 
> > Thanks @gabekassel. Can you provide a packet trace so that @hemanth-silabs and @suveshpratapa can debug it.
> 
> Attached, though not sure I'm capturing everything desired. This is wireshark using pyspinel setup with the network key. On the OTBR logging side, I just see continual beacon requests.
> 
> Screen Shot 2022-05-27 at 12 40 19 PM
> 
> [beacon_payload.pcapng.zip](https://github.com/openthread/openthread/files/8789378/beacon_payload.pcapng.zip)

Thanks again @gabekassel - I'll defer to @suveshpratapa and @hemanth-silabs now to see if this is enough info to figure why the change did not solve the issue.

File: src/core/config/mac.h
Comment (hemanth-silabs) [reply src/core/config/mac.h:510]: > > > > > @gabekassel it would be great if you could try out the changes in this PR and verify that it fixes the issue you are experiencing.
> > > > 
> > > > 
> > > > @romacdon I built [this](https://github.com/openthread/ot-br-posix/compare/main...gabekassel:gkass/beacon-payload-revert?expand=1) with `-DOPENTHREAD_CONFIG_MAC_BEACON_PAYLOAD_ENABLE=1` defined in my cmake args and still see failure with the commercial device in question.
> > > 
> > > 
> > > Thanks @gabekassel. Can you provide a packet trace so that @hemanth-silabs and @suveshpratapa can debug it.
> > 
> > 
> > Attached, though not sure I'm capturing everything desired. This is wireshark using pyspinel setup with the network key. On the OTBR logging side, I just see continual beacon requests.
> > Screen Shot 2022-05-27 at 12 40 19 PM
> > [beacon_payload.pcapng.zip](https://github.com/openthread/openthread/files/8789378/beacon_payload.pcapng.zip)
> 
> Thanks again @gabekassel - I'll defer to @suveshpratapa and @hemanth-silabs now to see if this is enough info to figure why the change did not solve the issue.



> > > > > @gabekassel it would be great if you could try out the changes in this PR and verify that it fixes the issue you are experiencing.
> > > > 
> > > > 
> > > > @romacdon I built [this](https://github.com/openthread/ot-br-posix/compare/main...gabekassel:gkass/beacon-payload-revert?expand=1) with `-DOPENTHREAD_CONFIG_MAC_BEACON_PAYLOAD_ENABLE=1` defined in my cmake args and still see failure with the commercial device in question.
> > > 
> > > 
> > > Thanks @gabekassel. Can you provide a packet trace so that @hemanth-silabs and @suveshpratapa can debug it.
> > 
> > 
> > Attached, though not sure I'm capturing everything desired. This is wireshark using pyspinel setup with the network key. On the OTBR logging side, I just see continual beacon requests.
> > Screen Shot 2022-05-27 at 12 40 19 PM
> > [beacon_payload.pcapng.zip](https://github.com/openthread/openthread/files/8789378/beacon_payload.pcapng.zip)
> 
> Thanks again @gabekassel - I'll defer to @suveshpratapa and @hemanth-silabs now to see if this is enough info to figure why the change did not solve the issue.

Thanks @gabekassel for the logs. It looks like there is no beacon at all. Also all the packet seems to have CRC errors. Maybe the beacon requests are getting dropped due to CRC errors?

This fix comes into picture if the beacon requests are processed and beacons are sent out. Looking at the logs it looks like the beacon requests are not being processed.

If this is on OTBR is it possible for you to collect some OTBR logs? Maybe we can see what happens to beacon requests and if they are getting dropped because of `Bad FCS`

Comment (hemanth-silabs) [reply src/core/config/mac.h:510]: > Yes, this should default to 1.

@romacdon @gabekassel @romacdon @edmont If we default the macro to 1, do we even need the macro? Can we get rid of it altogether?

Comment (gabekassel) [reply src/core/config/mac.h:510]: i do see beacons going out from the router. logs attached
[eero.log](https://github.com/openthread/openthread/files/8789580/eero.log)

Comment (jwhui) [reply src/core/config/mac.h:510]: > @romacdon @gabekassel @romacdon @edmont If we default the macro to 1, do we even need the macro? Can we get rid of it altogether? Especially since with default configuration, we will see lot of issues with older devices.

Removing the configuration to transmit beacons makes sense to me. I would keep the configuration for receiving the beacon payload and default to off to discourage its use going forward.

Comment (hemanth-silabs) [reply src/core/config/mac.h:510]: @gabekassel I am not sure what is going on here. The logs do show the beacons going out, but I don't see the beacons in the OTA capture. Can you let me know what commercial device this is?

Comment (gabekassel) [reply src/core/config/mac.h:510]: I've tested Nanoleaf Essential, Eve Energy, and Wemo Plug. All behave the same.

**Comment by hemanth-silabs** [comment]
@edmont @abtink Can you guys please review this PR as well?

**Comment by gabekassel** [comment]
Silabs team and I did some live testing today, and these changes resolve the interop issues I've observed in ToT before these changes. cc: @jwhui @ndyck14

**Comment by hemanth-silabs** [comment]
> Silabs team and I did some live testing today, and these changes resolve the interop issues I've observed in ToT before these changes. cc: @jwhui @ndyck14

Thanks @gabekassel for testing this again!!

**Review by romacdon — APPROVED** [review]
LGTM

**Review by suveshpratapa — APPROVED** [review]
LGTM

**Comment by romacdon** [comment]
@edmont @jwhui @gabekassel @abtink - I know this may be asking a lot but it would help the Silabs folks meet some internal deliverables if we could get this reviewed and merged today.  I'm hoping the successful testing we did with Gabe gives you confidence in the changes as you review the PR.

File: src/core/mac/mac_frame.hpp
Comment (abtink) [inline src/core/mac/mac_frame.hpp:1570]: Suggest instead of defining and using `kMaxSize` in `BeaconPayload` we directly use the definition in `NetworkName` (makes it clear that it is max size of `NetworkName` and avoid re-defining same constant).

```c++
char mNetworkName[MeshCop::NetworkName::kMaxSize];
```

File: src/core/mac/mac_frame.hpp
Comment (abtink) [inline src/core/mac/mac_frame.hpp:1606]: Checking the implementation before 
- https://github.com/openthread/openthread/pull/7472

we did use `NameData` here which would take care of all these situations (null termination, etc) and is safe to use.

Suggest using the same model as it was before. Here is the link to the code right before the PR above was merged:

https://github.com/openthread/openthread/blob/a8b4fbb1893823e1c247ee1cc7dac8f61519cba4/src/core/mac/mac_frame.hpp#L1593-L1638

**Review by abtink — COMMENTED** [review]
The change in this PR look fine.

I noticed one possible situation with `mNetworkName` field in `BeaconPayload`. I think it may not be null terminated in all cases (if we have max length network name). Added some suggestions below:

From what I recall, the "network name" field in a beacon may not be null terminated in all cases (in particular when the name is max length), so treating it as a C string may not be correct. I added suggesting below to use the same model as before (which uses `NameData` which would handle all these situations).

**Comment by hemanth-silabs** [comment]
> The change in this PR look fine.
> 
> I noticed one possible situation with `mNetworkName` field in `BeaconPayload`. I think it may not be null terminated in all cases (if we have max length network name) so treating it as a C string may not be correct in all cases.
> 
> Added some suggestions below:

@abtink I have reverted the changes that was using network name as c string and now it uses NameData as it was done before. Please have a look.

**Comment by hemanth-silabs** [comment]
@edmont Do you have any inputs on this PR?

**Comment by edmont** [comment]
> @edmont Do you have any inputs on this PR?

In general I'm okay with the change, still I'd prefer to have the beacon payload configurable even when defaults to 1, since it's an out of spec feature now.

**Comment by hemanth-silabs** [comment]
> > @edmont Do you have any inputs on this PR?
> 
> In general I'm okay with the change, still I'd prefer to have the beacon payload configurable even when defaults to 1, since it's an out of spec feature now.

@edmont I have added the beacon payload creation under a new macro which is enabled by default. Can you have a look please?

**Comment by hemanth-silabs** [comment]
@jwhui Looks like we have a consent here. @gabekassel yesterday confined the fix and @edmont okayed the changes with a request to provide a way to disable beacon payload if needed. I have updated the PR with a new configuration, that is enabled by default but can be disabled if needed.

Can you have a look and if the changes are OK, can you merge the PR?

**Comment by jwhui** [comment]
> Can you have a look and if the changes are OK, can you merge the PR?

@hemanth-silabs , it appears that some CI tests [are failing](https://github.com/openthread/openthread/runs/6694852065?check_suite_focus=true#step:6:4351).
```
======================================================================
ERROR: test (__main__.Cert_5_8_04_SecurityPolicyTLV)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/openthread/openthread/tests/scripts/thread-cert/thread_cert.py", line 304, in tearDown
    self._verify_packets(test_info_path)
  File "/home/runner/work/openthread/openthread/tests/scripts/thread-cert/thread_cert.py", line 333, in _verify_packets
    self.verify(pv)
  File "./tests/scripts/thread-cert/Cert_5_8_04_SecurityPolicyTLV.py", line 352, in verify
    pkts.filter_wpan_src64(LEADER).\
  File "/home/runner/work/openthread/openthread/tests/scripts/thread-cert/pktverify/packet_filter.py", line 234, in must_next
    raise errors.PacketNotFound(self.index, self._stop_index)
pktverify.errors.PacketNotFound: ((190, 190), (275, 275))

----------------------------------------------------------------------
```

I believe this test is checking whether the beacon payload is properly excluded when the B flag is not set in the Security Policy TLV. Of course, this PR includes this beacon payload without respecting the B flag. We could simply remove the check from the CI here. But the CI test implements Thread Certification Test 5.8.4, which includes a check that the Beacon payload is not included when the B flag is cleared. Unless the Thread Certification Test Specification is updated, I expect a similar failure in the Thread Test Harness as well. Thoughts?

**Comment by suveshpratapa** [comment]
> > Can you have a look and if the changes are OK, can you merge the PR?
> 
> @hemanth-silabs , it appears that some CI tests [are failing](https://github.com/openthread/openthread/runs/6694852065?check_suite_focus=true#step:6:4351).
> 
> ```
> ======================================================================
> ERROR: test (__main__.Cert_5_8_04_SecurityPolicyTLV)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/runner/work/openthread/openthread/tests/scripts/thread-cert/thread_cert.py", line 304, in tearDown
>     self._verify_packets(test_info_path)
>   File "/home/runner/work/openthread/openthread/tests/scripts/thread-cert/thread_cert.py", line 333, in _verify_packets
>     self.verify(pv)
>   File "./tests/scripts/thread-cert/Cert_5_8_04_SecurityPolicyTLV.py", line 352, in verify
>     pkts.filter_wpan_src64(LEADER).\
>   File "/home/runner/work/openthread/openthread/tests/scripts/thread-cert/pktverify/packet_filter.py", line 234, in must_next
>     raise errors.PacketNotFound(self.index, self._stop_index)
> pktverify.errors.PacketNotFound: ((190, 190), (275, 275))
> 
> ----------------------------------------------------------------------
> ```
> 
> I believe this test is checking whether the beacon payload is properly excluded when the B flag is not set in the Security Policy TLV. Of course, this PR includes this beacon payload without respecting the B flag. We could simply remove the check from the CI here. But the CI test implements Thread Certification Test 5.8.4, which includes a check that the Beacon payload is not included when the B flag is cleared. Unless the Thread Certification Test Specification is updated, I expect a similar failure in the Thread Test Harness as well. Thoughts?

@jwhui @hemanth-silabs Sounds like we should not enable `OPENTHREAD_CONFIG_MAC_BEACON_PAYLOAD_PARSING_ENABLE` for the test.

**Comment by hemanth-silabs** [comment]
> 

I agree. I think for all certification builds, we can and should safely turn this feature off. This should only be included for production devices to enable backward compatibility. Is there a macro that defines the build is for a test device? If so, maybe we can link both the macros in `mac.h`

**Comment by hemanth-silabs** [comment]
> I agree. I think for all certification builds, we can and should safely turn this feature off. This should only be included for production devices to enable backward compatibility. Is there a macro that defines the build is for a test device? If so, maybe we can link both the macros in `mac.h`

@jwhui @suveshpratapa What if we disable the beacon payload if `OT_REFERENCE_DEVICE` is defined?

**Comment by suveshpratapa** [comment]
> > I agree. I think for all certification builds, we can and should safely turn this feature off. This should only be included for production devices to enable backward compatibility. Is there a macro that defines the build is for a test device? If so, maybe we can link both the macros in `mac.h`
> 
> @jwhui @suveshpratapa What if we disable the beacon payload if `OT_REFERENCE_DEVICE` is defined?

This sounds good to me.

**Comment by jwhui** [comment]
> I agree. I think for all certification builds, we can and should safely turn this feature off. This should only be included for production devices to enable backward compatibility. Is there a macro that defines the build is for a test device? If so, maybe we can link both the macros in `mac.h`

To clarify, I believe any DUT with this feature enabled will currently fail Thread Certification with the current Thread Test Specification.

**Comment by hemanth-silabs** [comment]
> > I agree. I think for all certification builds, we can and should safely turn this feature off. This should only be included for production devices to enable backward compatibility. Is there a macro that defines the build is for a test device? If so, maybe we can link both the macros in `mac.h`
> 
> To clarify, I believe any DUT with this feature enabled will currently fail Thread Certification with the current Thread Test Specification.

The only way out then is just disable it by default and enable it in specific applications @romacdon @gabekassel.

**Comment by hemanth-silabs** [comment]
@jwhui @suveshpratapa @romacdon @gabekassel I have disabled including beacon payload by default. And letting the applications enable the feature as necessary. This is due to the fact that it impacts thread certification tests.

File: include/openthread/instance.h
Comment (jwhui) [inline include/openthread/instance.h:56]: I don't see any change to the API. Can we avoid this change?

Comment (hemanth-silabs) [reply include/openthread/instance.h:56]: I agree. I will revert back the version number bump.

**Review by jwhui — APPROVED** [review]

**Comment by edmont** [comment]
Sorry, I'm late to the discussion, but I agree with the changes.

**Comment by hemanth-silabs** [comment]
> Sorry, I'm late to the discussion, but I agree with the changes.

No issues @edmont and thank you for your feedback!!

**Pull request merged**
